src/zarr/storage/_cache.py

```diff
@@ -0,0 +1,698 @@
+                    result = key in cast(dict[str, Any], self._store)
+
+                self._contains_cache[key] = bool(result)
+            return bool(self._contains_cache[key])
+
+    async def clear(self) -> None:
+        """
+        Remove all keys from the store and clear the cache.
+
+        This operation clears both the underlying store and invalidates
+        all cached data to maintain consistency.
+        """
+        # Check if store is writable
+        self._check_writable()
+
+        await self._store.clear()
+        self.invalidate()
+
+    def keys(self) -> Iterator[str]:
+        """
+        Return an iterator over the keys in the store.
+        """
+        with self._mutex:
+            return iter(self._keys())
+
+    def _keys(self) -> list[str]:
+        if self._keys_cache is None:
+            # Check if it's a Store object vs dict-like object
+            if hasattr(self._store, "supports_listing"):
+                # It's a Store object
+                if self._store.supports_listing:
+                    try:
+
+                        async def collect_keys() -> list[str]:
+                            return [str(key) async for key in self._store.list()]
+
+                        keys = asyncio.run(collect_keys())
+                        self._keys_cache = keys
+                    except RuntimeError:
+                        # Already in async context
+                        raise NotImplementedError(
+                            "Cannot list keys in async context - _keys() method needs to be async"
+                        ) from None
+                else:
+                    # Store doesn't support listing
+                    self._keys_cache = []
+            else:
+                # It's a dict-like object (for tests)
+                if hasattr(self._store, "keys") and callable(self._store.keys):
+                    keys = [str(k) for k in self._store.keys()]  # noqa: SIM118
+                    self._keys_cache = keys
+                else:
+                    self._keys_cache = []
+        return self._keys_cache
+
+    def listdir(self, path: Path) -> list[str]:
+        """
+        Return a list of directory contents for the given path with caching.
+
+        This method provides a cached version of directory listings to improve
+        performance for repeated directory access operations.
+        """
+        with self._mutex:
+            # Normalize path to string for consistent caching
+            path_key = self._normalize_key(path)
+            try:
+                return self._listdir_cache[path_key]
+            except KeyError:
+                listing = listdir(self._store, path)
+                self._listdir_cache[path_key] = listing
+                return listing
+
+    async def getsize(self, key: str) -> int:
+        """
+        Get the size in bytes of the value stored at the given key.
+        """
+        return await self._store.getsize(key)
```

Comment:
Shouldn't you be checking if the value is in the cache, and if it is getting the size from the value in the cache, instead of going into the underlying store every time?